packages/workspace/src/node/default-workspace-server.spec.ts

```diff
@@ -0,0 +1,123 @@
+                }).trim();
+            }
+
+            const module = new ContainerModule(bind => {
+                bind(WorkspaceCliContribution).toSelf().inSingletonScope();
+                bind(CommonWorkspaceUtils).toSelf().inSingletonScope();
+                bind(DefaultWorkspaceServer).toSelf().inSingletonScope();
+                bind(EnvVariablesServer).toConstantValue(new MockEnvVariablesServerImpl(FileUri.create(fs.realpathSync(tempDirPath))));
+            });
+
+            container.load(module);
+
+            cut = container.get(DefaultWorkspaceServer);
+            envServerMock = container.get(EnvVariablesServer);
+        });
+
+        it('should return empty list of workspaces if no recent workspaces file is existing', async function (): Promise<void> {
+            const recent = await cut.getRecentWorkspaces();
+            expect(recent).to.be.empty;
+        });
+
+        it('should not return non-existing workspaces from recent workspaces file', async function (): Promise<void> {
+            const configDirUri = await envServerMock.getConfigDirUri();
+            const filePath = path.resolve(FileUri.fsPath(configDirUri), 'recentworkspace.json');
+
+            fs.writeFileSync(filePath, JSON.stringify({
+                recentRoots: [
+                    '/home/some/thing',
+                    '/home/some/other/thing'
+                ]
+            }));
+
+            const recent = await cut.getRecentWorkspaces();
+
+            expect(recent).to.be.empty;
+        });
+
+        it('should return only existing workspaces from recent workspaces file', async function (): Promise<void> {
+            const configDirUri = await envServerMock.getConfigDirUri();
+            let dirPath = FileUri.fsPath(configDirUri);
+            const filePath = path.resolve(dirPath, 'recentworkspace.json');
+
+            if (process.platform === 'win32') {
+                dirPath = new URI('file:///' + dirPath.toString()).toString();
+            }
+
+            fs.writeFileSync(filePath, JSON.stringify({
+                recentRoots: [
+                    dirPath.toString(),
+                    '/home/some/other/thing'
+                ]
+            }));
+
+            const recent = await cut.getRecentWorkspaces();
+
+            expect(recent).to.have.members([dirPath.toString()]);
+        });
+
+        it('should return empty list of workspaces if recent workspaces file is malformed', async function (): Promise<void> {
+            const configDirUri = await envServerMock.getConfigDirUri();
+            const filePath = path.resolve(FileUri.fsPath(configDirUri), 'recentworkspace.json');
+
+            // previously caused: 'TypeError: Cannot read property 'fsPath' of undefined'
+            fs.writeFileSync(filePath, `{
+                "recentRoots": [
+                  [
+                    "file:///home/some/path"
+                  ],
+                  "/home/some/other/path"
+                ]
```

Comment:
I don't remember the details but this type of corruption rings a bell. Did it happen again in your environment?

I was thinking we could implement some kind of "best effort" when parsing this data:

```ts
namespace RecentWorkspacePathsData {
    export function is(data: unknown): data is RecentWorkspacePathsData {
        // eslint-disable-next-line no-null/no-null
        return data !== null
            && typeof data === 'object'
            && Array.isArray((data as RecentWorkspacePathsData).recentRoots)
            && (data as RecentWorkspacePathsData).recentRoots.every(root => typeof root === 'string');
    }
    /**
     * Try to parse `data` as `RecentWorkspacePathsData` but will remove incompatible data.
     */
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    export function cleanup(data: any): RecentWorkspacePathsData | undefined {
        // eslint-disable-next-line no-null/no-null
        if (typeof data !== 'object' || data === null || !Array.isArray(data.recentRoots)) {
            return;
        }
        return {
            recentRoots: data.recentRoots.filter((root: unknown) => typeof root === 'string')
        };
    }
}
```

Then update `readRecentWorkspacePathsFromUserHome`:

```ts
    protected async readRecentWorkspacePathsFromUserHome(): Promise<RecentWorkspacePathsData | undefined> {
        const fsPath = await this.getUserStoragePath();
        const data = await this.readJsonFromFile(fsPath);
        return RecentWorkspacePathsData.cleanup(data);
    }
```

This way if we get simple corruptions like weird elements we can still recover the rest. WDYT?

Reply:
sounds good to me, adapted as you suggested :+1: 
I wasn't experiencing this issue with theia.